src/haxelib/api/Vcs.hx

```diff
@@ -188,30 +164,93 @@ class Vcs implements IVcs {
 		return available;
 	}
 
-	@:final function get_available():Bool {
+	final function get_available():Bool {
 		if (!availabilityChecked)
 			checkExecutable();
 		return available;
 	}
 
-	public function clone(libPath:String, vcsPath:String, ?branch:String, ?version:String):Void {
-		throw "This method must be overridden.";
+	final function runStrict(args:Array<String>, ?debugLog:(msg:String) -> Void):{
+		code:Int,
+		out:String,
+		err:String
+	} {
+		inline function print(msg)
+			if (debugLog != null && msg != "")
+				debugLog(msg);
+
+		print("# Running command: " + executable + " " + args.toString() + "\n");
+
+		final proc = command(executable, args);
+		switch (proc) {
+			case {code: 0}: // pass
+				print(proc.out);
+				print(proc.err);
+				print('# Exited with code ${proc.code}\n');
+			case {code: code, out: out, err: err}:
+				throw CommandFailed(this, code, out, err);
+		}
+		return proc;
+	}
+
+	final function run(args:Array<String>, ?debugLog:(msg:String) -> Void):{
+		code:Int,
+		out:String,
+		err:String
+	} {
+		inline function print(msg)
+			if (debugLog != null && msg != "")
+				debugLog(msg);
+
+		print("# Running command: " + executable + " " + args.toString() + "\n");
+
+		final proc = command(executable, args);
+
+		print(proc.out);
+		print(proc.err);
+		print('# Exited with code ${proc.code}\n');
+
+		return proc;
 	}
 
-	public function update(libName:String):Bool {
-		throw "This method must be overridden.";
+	static function command(cmd:String, args:Array<String>) {
+		final p = try {
+			new sys.io.Process(cmd, args);
+		} catch (e:Dynamic) {
+			return {
+				code: -1,
+				out: "",
+				err: Std.string(e)
+			}
+		}
+		final out = p.stdout.readAll().toString();
```

Comment:
For the future: This is not guaranteed to work because stderr could be full before stdout finishes reading and vice versa. This is actually a really annoying problem with our Process API and the only way around it is to use threads to make sure the pipes don't clog. We should at some point look into this situation, even if in practice it is unlikely to make much of a difference.

Reply:
Would it be fixed if the exitCode() call comes before those two?

Reply:
No, the process still won't terminate if its pipes are full. Calling exitCode would trigger a flush, but that doesn't help if nobody picks anything up on the other side.